storage/blockchain/src/service/read.rs

```diff
@@ -32,156 +29,34 @@ use crate::{
     },
     service::{
         free::{compact_history_genesis_not_included, compact_history_index_to_height_offset},
-        types::{ResponseReceiver, ResponseResult, ResponseSender},
+        types::{BCReadHandle, ResponseResult},
     },
     tables::OpenTables,
     tables::{BlockHeights, BlockInfos, Tables},
     types::{Amount, AmountIndex, BlockHash, BlockHeight, KeyImage, PreRctOutputId},
 };
 
-//---------------------------------------------------------------------------------------------------- DatabaseReadHandle
-/// Read handle to the database.
+//---------------------------------------------------------------------------------------------------- init_read_service
+/// Initialize the [`BCReadHandle`] thread-pool backed by `rayon`.
 ///
-/// This is cheaply [`Clone`]able handle that
-/// allows `async`hronously reading from the database.
+/// This spawns `threads` amount of reader threads
+/// attached to `env` and returns a handle to the pool.
 ///
-/// Calling [`tower::Service::call`] with a [`DatabaseReadHandle`] & [`BCReadRequest`]
-/// will return an `async`hronous channel that can be `.await`ed upon
-/// to receive the corresponding [`BCResponse`].
-pub struct DatabaseReadHandle {
-    /// Handle to the custom `rayon` DB reader thread-pool.
-    ///
-    /// Requests are [`rayon::ThreadPool::spawn`]ed in this thread-pool,
-    /// and responses are returned via a channel we (the caller) provide.
-    pool: Arc<rayon::ThreadPool>,
-
-    /// Counting semaphore asynchronous permit for database access.
-    /// Each [`tower::Service::poll_ready`] will acquire a permit
-    /// before actually sending a request to the `rayon` DB threadpool.
-    semaphore: PollSemaphore,
-
-    /// An owned permit.
-    /// This will be set to [`Some`] in `poll_ready()` when we successfully acquire
-    /// the permit, and will be [`Option::take()`]n after `tower::Service::call()` is called.
-    ///
-    /// The actual permit will be dropped _after_ the rayon DB thread has finished
-    /// the request, i.e., after [`map_request()`] finishes.
-    permit: Option<OwnedSemaphorePermit>,
-
-    /// Access to the database.
-    env: Arc<ConcreteEnv>,
+/// Should be called _once_ per actual database.
+#[cold]
+#[inline(never)] // Only called once.
+pub fn init_read_service(env: Arc<ConcreteEnv>, threads: ReaderThreads) -> BCReadHandle {
```

Comment:
This is a public function now so this should be behind `# Invariant` and it should explain what happens if it is broken.

Other functions too:
- `init_read_service_with_pool`
- `DatabaseReadService::new`

Reply:
I added some more docs here, however I don't think anything bad will happen if this is called multiple times right?